--- ngx_cache_purge_module.c    2014-12-23 18:40:21.000000000 +0000
+++ ngx_cache_purge_module.c.new        2017-05-04 00:11:09.337073629 +0000
@@ -492,7 +492,11 @@
     ngx_str_t                      body_source;
 #  endif /* nginx_version < 1007008 */
 
+#  if (nginx_version >= 1011006)
+    ngx_http_complex_value_t      *method;
+#  else
     ngx_str_t                      method;
+#  endif /* nginx_version >= 1011006 */
     ngx_str_t                      location;
     ngx_str_t                      url;
